packages/typespec-client-generator-core/src/validate.ts

```diff
@@ -1,12 +1,205 @@
-import { Program } from "@typespec/compiler";
-import { createTCGCContext } from "./internal-utils.js";
-import { getAllModelsWithDiagnostics } from "./types.js";
+import {
+  AugmentDecoratorStatementNode,
+  DecoratorExpressionNode,
+  Enum,
+  EnumMember,
+  Interface,
+  Model,
+  ModelProperty,
+  Namespace,
+  Operation,
+  Program,
+  projectProgram,
+  Scalar,
+  Type,
+  Union,
+  UnionVariant,
+} from "@typespec/compiler";
+import { DuplicateTracker } from "@typespec/compiler/utils";
+import { buildVersionProjections, getVersions } from "@typespec/versioning";
+import { getClientNameOverride } from "./decorators.js";
+import { AllScopes, clientNameKey, createTCGCContext, TCGCContext } from "./internal-utils.js";
+import { reportDiagnostic } from "./lib.js";
 
 export function $onValidate(program: Program) {
-  // Pass along any diagnostics that might be returned from the HTTP library
   const tcgcContext = createTCGCContext(program);
-  const [_, diagnostics] = getAllModelsWithDiagnostics(tcgcContext);
-  if (diagnostics.length > 0) {
-    program.reportDiagnostics(diagnostics);
+  const languageScopes = getDefinedLanguageScopes(program);
+  for (const scope of languageScopes) {
+    validateClientNamesPerNamespace(program, tcgcContext, scope, program.getGlobalNamespaceType());
+  }
+}
+
+function getDefinedLanguageScopes(program: Program): Set<string | typeof AllScopes> {
+  const languageScopes = new Set<string | typeof AllScopes>();
+  for (const value of program.stateMap(clientNameKey).values()) {
+    if (value[AllScopes]) {
+      languageScopes.add(AllScopes);
+    }
+    for (const languageScope of Object.keys(value)) {
+      languageScopes.add(languageScope);
+    }
+  }
+  return languageScopes;
+}
+
+function validateClientNamesPerNamespace(
+  program: Program,
+  tcgcContext: TCGCContext,
+  scope: string | typeof AllScopes,
+  namespace: Namespace
+) {
+  const apiVersions = getVersions(program, namespace)[1]
+    ?.getVersions()
+    ?.map((x) => x.value);
+  if (apiVersions !== undefined) {
+    const versionProjections = buildVersionProjections(program, namespace);
```

Comment:
The versioning should be done differently this is extremely clostly it will project the whole program for every version and do that again in the emitters.  Reminder that all those things need to be able to run very fast to get IDE support

It also will cause duplicate failures [See playground](https://cadlplayground.z22.web.core.windows.net/typespec-azure/prs/1119/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZ8wfYXp1cmUtdG9vbHMvyCstxhVjb3Jl3ytjLWNsaWVudC1nZW5lcmF0b3LINgp1c2luZyBBxDQuQ8UlR8gkLkNvcmU7xyJUeXBlU3BlYy5IdHRw0RVSZXN00RVW6QDKyBvHZ8tXyhIuVHJhaXRzOwoKQHNlcnZpY2UKQOcBDGVkKENvbnRvc28uV2lkZ2V0TWFuYWdlcshicykKbmFtZXNwYWNlINUqOwoKLyoqIFRoySAgxiAgxyEgx3Qgx3MuICovCmVudW0gyGQgewogIHYxLMQGMiwKfQoKbW9kZWwgRm9vIHt9CgpA5gFjTmFtZSgiRm9vIinHIUJhcsQh&e=%40azure-tools%2Ftypespec-autorest&options=%7B%22linterRuleSet%22%3A%7B%22extends%22%3A%5B%22%40azure-tools%2Ftypespec-azure-rulesets%2Fdata-plane%22%5D%7D%7D)

It would be better to ignore the versioning case for now if its too complex than doing this.

Reply:
this is also probably not working when involving multiple service getting versioned individually.

In the versioning library all the validation is done by checking the data directly and checking if that name changed for example instead of traversing `n` graphs.

Reply:
> this is also probably not working when involving multiple service getting versioned individually.

I don't imagine how it can't be handled, but I agree we should skip it for now if projecting the whole program is extremely costly. Reverted the change for versioning.

> In the versioning library all the validation is done by checking the data directly and checking if that name changed for example instead of traversing n graphs

Just out of curiosity, could you point me to the place where versioning library is checking this?

Reply:
Versioning validation here https://github.com/microsoft/typespec/blob/main/packages/versioning/src/validate.ts

and not really as you can see here versioning is incredibly complex(and still full of bugs)